cpp/src/fil/infer.cu

```diff
@@ -260,21 +301,36 @@ struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_FEW_CLASSES> : finalize_block {
     }
     __syncthreads();  // free up per_thread[] margin
 
-    write_best_class_in_block(to_vec(threadIdx.x, acc), num_classes, out,
-                              num_rows);
+    int c = threadIdx.x;
+    if (num_outputs == 1) {  // will output class
+      write_best_class_in_block(to_vec(c, acc), num_classes, out, num_rows);
+    } else {  // output softmax-ed margin
+      // softmax phase 1: compute sum of exponents
+      auto max = block_all_max(acc, num_classes);
+      auto exponent = shifted_exp(acc, max);
+      auto soe = block_all_sum(exponent, num_classes);  // sum of exponents
+      // softmax phase 2: normalization happens while outputting
+      for (int i = 0; i < NITEMS; ++i) {
+        int row = blockIdx.x * NITEMS + i;
+        if (row < num_rows && c < num_classes) {
+          out[c + num_classes * row] = exponent[i] / soe[i];
+        }
+      }
+    }
   }
 };
 
 template <int NITEMS>
 struct tree_aggregator_t<NITEMS, GROVE_PER_CLASS_MANY_CLASSES>
   : finalize_block {
   vec<NITEMS, float> acc;
-  vec<NITEMS, float>* per_class_margin;
+  vec<NITEMS, float>* per_class;
```

Comment:
`per_class_value` could sounds like a better name. Could you also add a comment describing what it is used for?